core/server/common/src/main/java/alluxio/master/BackupManager.java

```diff
@@ -63,18 +80,87 @@ public BackupManager(MasterRegistry registry) {
    * @param os the stream to write to
    */
   public void backup(OutputStream os) throws IOException {
-    int count = 0;
+    // Create gZIP compressed stream as back-up stream.
```

Comment:
This method is getting pretty long. I think it could be cleaner to move this logic to a new stateful `BackupWriter`/`BackupReader` classes which handle a single calls to backup/restore. The tasks could be modeled with inner classes instead of anonymous functions. WDYT?

Reply:
Since both taking a backup and applying it revolve around a blocking queue and both use draining, I think there could be a shared base class that provides reader/writer infra. So then, splitted classes could be differing only on the actual operations they're doing. I would like to put a TODO and work on it later if that's fine.

Reply:
SG